src/Parser/Core/Modules/Items/BFA/Dungeons/BalefireBranch.js

```diff
@@ -0,0 +1,167 @@
+ * Taking damage removes stacks. It appears that taking larger hits removes more stacks.
+ * Stacks are removed even if the damage is 100% absorbed.
+ * 
+ * The initial buff shows in the log (and changebuffstack fabricated events) as a "1 stack" buff.
+ * The log shows the correct stack count once it has been lowered from that initial value.
+ */
+class BalefireBranch extends Analyzer {
+  static dependencies = {
+    abilities: Abilities,
+    spellUsable: SpellUsable,
+  }
+  applyCount = 0;
+  currentStack = 0;
+  totalUptime = 0;
+
+  lastApply = null;
+  lastChange = null;
+
+  // used to find average stacks over whole fight. 1 second at 5 stacks and 2 seconds at 10 would be sumStacks = 25
+  sumStacks = 0;
+
+  // sumStacks value if the player had taken no damage.
+  expectedSumStacks = 0;
+
+  // scales with trinket's ilvl
+  intellectPerStack = 0;
+  
+  constructor(...args) {
+    super(...args);
+    this.active = this.selectedCombatant.hasTrinket(ITEMS.BALEFIRE_BRANCH.id);
+
+    if (this.active) {
+      const itemLevel = this.selectedCombatant.getItem(ITEMS.BALEFIRE_BRANCH.id).itemLevel;
+      this.intellectPerStack = calculatePrimaryStat(BASE_ITEM_LEVEL, BASE_INTELLECT_PER_STACK, itemLevel);
+
+      // remind the player to activate the trinket
+      this.abilities.add({
+        spell: SPELLS.BALEFIRE_BRANCH_SPELL,
+        name: SPELLS.BALEFIRE_BRANCH_SPELL.name,
+        buffSpellId: SPELLS.KINDLED_SOUL.id,
+        category: Abilities.SPELL_CATEGORIES.ITEMS,
+        cooldown: ACTIVATION_COOLDOWN,
+        castEfficiency: {
+          suggestion: true,
+        },
+      });
+    }
+  }
+
+  on_toPlayer_applybuff(event) {
+    if (SPELLS.KINDLED_SOUL.id !== event.ability.guid) {
+      return;
+    }
+    this.stackChange(STACKS_START, event.timestamp);
+    this.expectedSumStacks += EXPECTED_SUM_STACKS_PER_USE;
+    this.applyCount += 1;
+    this.lastApply = event.timestamp;
+
+    // using the trinket doesn't trigger a cast event, so make our own
+    this.owner.fabricateEvent({
+      ...event,
+      ability: {
+        abilityIcon: SPELLS.BALEFIRE_BRANCH_SPELL.icon,
+        guid: SPELLS.BALEFIRE_BRANCH_SPELL.id,
+        name: SPELLS.BALEFIRE_BRANCH_SPELL.name,
+      },
+      type: 'cast',
+    }, event);
+  }
+
+  on_toPlayer_removebuffstack(event) {
+    if (SPELLS.KINDLED_SOUL.id !== event.ability.guid) {
+      return;
+    }
+    this.stackChange(event.stack, event.timestamp);
+  }
+
+  on_toPlayer_removebuff(event) {
+    if (SPELLS.KINDLED_SOUL.id !== event.ability.guid) {
+      return;
+    }
+    this.stackChange(0, event.timestamp);
+    this.totalUptime += (event.timestamp - this.lastApply);
+  }
+
+  on_finished() {
+    if (!this.lastChange || !this.currentStack) {
+      return;
+    }
+    const now = this.owner.fight.end_time;
+    this.stackChange(0, now);
+    this.totalUptime += (now - this.lastApply);
+
+    if (this.lastApply + DURATION_TO_DECAY > now) {
```

Comment:
nit: Using parenthesis around operators is preferred so it can be interpreted by humans without having to be aware of operator precedence.